Ensemble-Model.md

```diff
@@ -0,0 +1,1066 @@
+AMO_DUMP_BUCKET = 'telemetry-parquet'
+AMO_DUMP_KEY = 'telemetry-ml/addon_recommender/addons_database.json'
+```
+
+
+```python
+def load_amo_external_whitelist():
+    """ Download and parse the AMO add-on whitelist.
+    :raises RuntimeError: the AMO whitelist file cannot be downloaded or contains
+                          no valid add-ons.
+    """
+    final_whitelist = []
+    amo_dump = {}
+    try:
+        # Load the most current AMO dump JSON resource.
+        s3 = boto3.client('s3')
+        s3_contents = s3.get_object(Bucket=AMO_DUMP_BUCKET, Key=AMO_DUMP_KEY)
+        amo_dump = json.loads(s3_contents['Body'].read())
+    except ClientError:
+        logger.exception("Failed to download from S3", extra={
+            "bucket": AMO_DUMP_BUCKET,
+            "key": AMO_DUMP_KEY})
+
+    # If the load fails, we will have an empty whitelist, this may be problematic.
+    for key, value in amo_dump.items():
+        addon_files = value.get('current_version', {}).get('files', {})
+        # If any of the addon files are web_extensions compatible, it can be recommended.
+        if any([f.get("is_webextension", False) for f in addon_files]):
+            final_whitelist.append(value['guid'])
+
+    if len(final_whitelist) == 0:
+        raise RuntimeError("Empty AMO whitelist detected")
+
+    return final_whitelist
+```
+
+### Filtering out legacy addons
+
+The collaborative recommender should make recommendations based on WebExtensions. In general, our ensemble should also only recommend WebExtensions. Because of this, we need a way to filter for WebExtensions. We store this collection in two different data structures (list, set) as both will be useful in the remainder of the notebook.
+
+
+```python
+whitelist = load_amo_external_whitelist()
+whiteset = set(whitelist)
+```
+
+    INFO:botocore.vendored.requests.packages.urllib3.connectionpool:Starting new HTTP connection (1): 169.254.169.254
+    INFO:botocore.vendored.requests.packages.urllib3.connectionpool:Starting new HTTP connection (1): 169.254.169.254
+    INFO:botocore.vendored.requests.packages.urllib3.connectionpool:Starting new HTTPS connection (1): s3-us-west-2.amazonaws.com
+
+
+
+```python
+def get_whitelisted_addons(installed_addons):
+    return whiteset.intersection(installed_addons)
+```
+
+### Completing client data
+
+
+```python
+from dateutil.parser import parse as parse_date
+from datetime import datetime
+```
+
+
+```python
+def compute_weeks_ago(formatted_date):
+    try:
+        date = parse_date(formatted_date).replace(tzinfo=None)
+    except ValueError: # raised when the date is in an unknown format
+        return float("inf")
+    
+    days_ago = (datetime.today() - date).days
+    return days_ago / 7
+```
+
+
+```python
+def complete_client_data(client_data):
+    client = client_data.asDict()
+    
+    addons = client['installed_addons'] or []
+    client['installed_addons'] = get_whitelisted_addons(addons)
+    client['disabled_addons_ids'] = addons
```

Comment:
Why are you setting `disabled_addons_ids` to the `installed_addons`? I know that we don't have the disabled add-on info in the Longitudinal dataset. However, only putting legacy addons would be more appropriate here. Could you expand your reasoning here?

Reply:
I'm just using that dictionary key because it is the one that the [`LegacyRecommender`](https://github.com/mozilla/taar/blob/master/taar/recommenders/legacy_recommender.py#L44) expects. The LegacyRecommender only takes into account addons in its replacement list anyways, so adding another filter for legacy addons on my part wouldn't make a difference.